### ISSUE TITLE
Internal User Authentication

### ISSUE THREAD

**Issue by torrin47** [USEPA/EPA_Non-geo_Metadata_Editor#28]
_From @torrin47 on August 31, 2018 23:3_

Internal EPA users will have the ability to login to the site using the same EPA GeoPlatform SAML authentication the EDG is migrating to.

_Copied from original issue: USEPA/EPA_Environmental_Dataset_Gateway#61_

**Comment by aergul** [comment]
@torrin47 we've initially implemented using the REST API of the authentication service and we didn't get much more than the token and the username. Using ESRI libs, we get a bit further and are able to determine the user's full name as well as:
```
role: "org_publisher"
roleId: "jmc1ObdWfBTH6NAN"
```

I believe these indicate the user's role on the ArcGIS Online which most likely is unaware of the user's role in Geoportal Server. I could be wrong though, is there some mechanism that syncs or otherwise makes this info from EDG available to the EPA GeoPlatform?

If the answer is no, then I can imagine two alternatives:
1- We can make a call to EDG with the user's token to determine if the user is an EDG admin or publisher. We may need to erect a service for this if no API is readily available on the EDG side for this. 
2- The services we will build to load/save a record from/to EDG can perform the check at that time. Somewhat less than ideal as the user won't know that they can actually load/save until attempting to do so.

Preference?

**Comment by torrin47** [comment]
You're correct that the role and roleId are specific to ArcGIS Online and aren't related to EDG functionality. EDG permissions are controlled entirely by "group" membership - and I'm guessing that if you are receiving the role for the user, you also see the user's groups?  The groups listed above are the relevant ones - there are two master groups - publisher and admin - and then all the individual organizational groups.  The EDG allows a user to log in and then makes a server-side call for the user details and compares group membership against the list we embed in the config file. What I've never investigated is whether there's an API on the EDG side that we can use to check whether a logged in user is recognized as in the right GeoPlatform groups. It seems like the kind of thing we ought to be able to delegate to the EDG rather than have to check separately, but I've not seen any documentation on this front. I guess this is your alternative #1.  Does that help?

**Comment by aergul** [comment]
Well, the thing is we are authenticating against EPA GeoPlatform / ArcGIS Online which is not aware of EDG groups. We can worry about this when we get to #26 and #27.

This ticket has now been deployed to dev. To facilitate this, dirty checking has also been implemented to prevent navigation confirmation prompts convoluting authentication related navigation such as: 

![image.png](https://images.zenhubusercontent.com/5b888c67b1ed9a7d4f8504bf/7222a117-2279-4108-9030-e058604ffca4)

There may be some edge cases around dirty checks so emphasis will need to be on QA @jzichichi @torrin47 .

**Comment by jzichichi** [comment]
@aergul @torrin47 this worked well for me; I tested with login, without login, etc.  The prompts all worked as I would have expected.

**Comment by torrin47** [comment]
Basic Login/logout works well.  A couple of notes:

- Login option only displays on pageload. If a user initially continues without login or logs in and logs out, there's no way to trigger the login other than reloading the page. Can we add a Login button to either the title bar (where logout appears) or to the left-hand Load/View/Save menu?
- EPA Contact/sponsor disappears after login, but EPA Agreement is still visible - this should also disappear.

**Comment by aergul** [comment]
EPA Contact/sponsor addressed.

Login/logout flow is being implemented under #71
